src/Zigzag_persistence/include/gudhi/zigzag_persistence.h

```diff
@@ -0,0 +1,451 @@
+      births_.erase(chainFp);
+    }
+
+    // Update persistence diagram with left interval [fil(b_max) ; fil(m))
+    stream_interval_(dim - 1, maxb, numArrow_);
+  }
+
+  /**
+   * @brief Removes the given face by pushing up the matrix the corresponding column and erasing it.
+   *
+   * @param faceID Internal ID of the face to remove.
+   * @param dim Dimension of the face to remove.
+   */
+  void _process_backward_arrow(index faceID, dimension_type dim) {
+    // column whose key is the one of the removed face
+    matrix_index currCol = matrix_.get_column_with_pivot(faceID);
+
+    // Record all columns that get affected by the transpositions, i.e., have a coeff
+    std::vector<matrix_index> modifiedColumns;
+    const auto& row = matrix_.get_row(faceID);
+    modifiedColumns.reserve(row.size());
+    std::transform(row.begin(), row.end(), std::back_inserter(modifiedColumns),
+                   [](const auto& cell) { return cell.get_column_index(); });
+    // Sort by left-to-right order in the matrix_ (no order maintained in rows)
+    std::stable_sort(modifiedColumns.begin(), modifiedColumns.end(), [this](matrix_index i1, matrix_index i2) {
+      return matrix_.get_pivot(i1) < matrix_.get_pivot(i2);
+    });
+
+    // Modifies curr_col, not the other one.
+    for (auto otherColIt = std::next(modifiedColumns.begin()); otherColIt != modifiedColumns.end(); ++otherColIt) {
+      currCol = matrix_.vine_swap_with_z_eq_1_case(currCol, *otherColIt);
+    }
+
+    // curr_col points to the column to remove by restriction of K to K-{\sigma}
+    if (!matrix_.get_column(currCol).is_paired()) {  // in F
+      auto it = births_.find(currCol);
+      stream_interval_(dim, it->second, numArrow_);
+      if constexpr (erase_birth_history) {
+        birthOrdering_.remove_birth(it->second);
+        births_.erase(it);
+      }
+    } else {  // in H    -> paired with c_g, that now belongs to F now
+      // maintain the <=b order
+      birthOrdering_.add_birth_backward(numArrow_);
+      if constexpr (erase_birth_history) {
+        births_.try_emplace(matrix_.get_column(currCol).get_paired_chain_index(), numArrow_);
+      } else {
+        auto res = births_.try_emplace(matrix_.get_column(currCol).get_paired_chain_index(), numArrow_);
+        if (!res.second) {
+          res.first->second = numArrow_;
+        }
+      }
+    }
+
+    // cannot be in G as the removed face is maximal
+    matrix_.remove_maximal_face(faceID, {});  // also unpaires c_g if in H
+  }
+
+ private:
+  Matrix_type matrix_;           /**< Matrix storing a base of the current chain complex. */
+  birth_dictionnary births_;     /**< Map face index in F to corresponding birth. */
+  Birth_ordering birthOrdering_; /**< Maintains <b ordering of the births. */
+  index numArrow_;               /**< Current arrow number. */
+  std::function<void(dimension_type, index, index)> stream_interval_; /**< Callback method for closed pairs. */
```

Comment:
If `std::function` turns out to be costly (probably not?), an alternative would be `SI stream_interval_` where SI is an extra template parameter to the class. And since the other template parameters probably prevent from using CTAD, construction would be through a helper function to avoid having to spell out the type of the functor. ` auto zz = make_zigzag_persistence<options>(stream_interval);`.